K2Bridge/Utils/DateMathParser.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+namespace K2Bridge.Visitors
+{
+    using System.Text.RegularExpressions;
+
+    /// <content>
+    /// Parser for the Date Math expressions.
+    /// </content>
+    public class DateMathParser
+    {
+        public static string ParseDateMath(string expr)
+        {
+            // See Wiki or ES docs for details on date math syntax
+            var pattern = @"^(?<anchor>now|.+?(?:\|\||$))(?<ranges>(?:(?:\+|\-)[^\/]*))?(?<rounding>\/(?:y|M|w|d|h|H|m|s))?$";
```

Comment:
I don't really like this regex. It makes you do a lot of work to get the actual values from it.

If you want to keep it because we took it from somewhere else, that's ok (just do the comments in the rest of the file).
But I've took the liberty to rewrite this file to make it (in my opinion) nicer and more readable:

```c#
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

namespace K2Bridge.Visitors
{
    using System.Text.RegularExpressions;

    /// <content>
    /// Parser for the Date Math expressions.
    /// </content>
    public class DateMathParser
    {
        public static string ParseDateMath(string expr)
        {
            // See Wiki or ES docs for details on date math syntax
            var rgx = new Regex(
                @"^
                        (?: (?<anchor>now) | (?: (?<anchor>.+?) (?:\|\||$)) )
                        (?: (?<rangeSign>[+-]) (?<rangeValue>\d+) (?<rangeUnit>[a-zA-Z]))?
                        (?: (?:\/ (?<rounding>y|M|w|d|h|H|m|s)))?", RegexOptions.IgnorePatternWhitespace);
            var match = rgx.Match(expr);

            var kexpr = string.Empty;

            if (!match.Success)
            {
                return kexpr;
            }

            // The date literal, ie.g. YYYY-MM-DD or "now"
            var anchor = match.Groups["anchor"].Value;
            if (anchor.Equals("now"))
            {
                kexpr = "now()";
            }
            else
            {
                kexpr = $"make_datetime('{anchor}')";
            }

            // Date operation, e.g. -1M, +3d, etc.
            if (match.Groups["rangeSign"].Success)
            {
                var sign = match.Groups["rangeSign"].Value == "-" ? "-" : string.Empty;
                var value = int.Parse(match.Groups["rangeValue"].Value);
                var unit = match.Groups["rangeUnit"].Value[0] switch {
                    'y' => "year",
                    'M' => "month",
                    'w' => "week",
                    'd' => "day",
                    'h' => "hour",
                    'H' => "hour",
                    'm' => "minute",
                    's' => "second",
                    _ => null,
                };

                if (unit != null)
                {
                    kexpr = $"datetime_add('{unit}', {sign}{value}, {kexpr})";
                }
            }

            // Rounding, e.g. /d, /h, /m, etc.
            if (match.Groups["rounding"].Success)
            {
                kexpr = match.Groups["rounding"].Value switch {
                    "y" => $"startofyear({kexpr})",
                    "M" => $"startofmonth({kexpr})",
                    "w" => $"startofweek({kexpr})",
                    "d" => $"startofday({kexpr})",
                    "h" => $"bin({kexpr}, 1h)",
                    "H" => $"bin({kexpr}, 1h)",
                    "m" => $"bin({kexpr}, 1m)",
                    "s" => $"bin({kexpr}, 1s)",
                    _ => kexpr,
                };
            }

            return kexpr;
        }
    }
}

```

So tell me what you think, I'm fine with either

Reply:
Thanks Asaf, your version looks cleaner.